### PR TITLE
llvm: test `scan-build`.

### PR DESCRIPTION
**Pull request by carlocab**
I've split this off from #106925 to see if `scan-build` failing will
block everything else again.

I've set this up to fail quickly everywhere but on ARM Monterey.

### CODE REVIEW

**Comment by carlocab** [comment]
Huh, seems to work now. Interesting.

File: Formula/llvm.rb [hunk 1]
```diff
@@ -33,25 +32,36 @@ class Llvm < Formula
   depends_on "cmake" => :build
   depends_on "swig" => :build
   depends_on "python@3.10"
+  depends_on "z3"
+  depends_on "zstd"
```
Comment (carlocab) [inline Formula/llvm.rb:36]:
```suggestion
```

File: Formula/llvm.rb [hunk 2]
```diff
@@ -102,15 +112,15 @@ def install
       -DLLVM_INCLUDE_DOCS=OFF
       -DLLVM_INCLUDE_TESTS=OFF
       -DLLVM_INSTALL_UTILS=ON
-      -DLLVM_ENABLE_Z3_SOLVER=OFF
+      -DLLVM_ENABLE_Z3_SOLVER=ON
```
Comment (carlocab) [inline Formula/llvm.rb:115]:
```suggestion
      -DLLVM_ENABLE_Z3_SOLVER=OFF
```

File: Formula/llvm.rb [hunk 3]
```diff
@@ -598,8 +637,8 @@ def caveats
         return 0;
       }
     EOS
-    assert_includes shell_output("#{bin}/scan-build clang++ scanbuildtest.cpp 2>&1"),
-      "warning: Use of memory after it is freed"
+    assert_includes shell_output("#{bin}/scan-build make scanbuildtest 2>&1"),
```
Comment (carlocab) [inline Formula/llvm.rb:640]:
```suggestion
    assert_includes shell_output("#{bin}/scan-build clang++ scanbuildtest.cpp 2>&1"),
```

**Comment by carlocab** [comment]
Here we go: https://github.com/Homebrew/homebrew-core/actions/runs/3053652971/jobs/4924662506#step:6:212

Invoking `scan-build clang++` errors, but `scan-build make` does not seem to.

Though, I still don't understand why `scan-build clang++` works for me locally.

**Comment by cho-m** [comment]
Failure is happening on changes in https://github.com/llvm/llvm-project/commit/157bbe6aea22e87c822f6cda3cd404b8f657dce4 / https://reviews.llvm.org/D114786

Though not sure why.

Is it mixing up Xcode and LLVM headers somewhere? Could try removing `CPATH`.

**Comment by carlocab** [comment]
> Is it mixing up Xcode and LLVM headers somewhere? Could try removing `CPATH`.

That would explain why I'm not seeing it locally, since I don't have the CLT.

**Comment by carlocab** [comment]
> Is it mixing up Xcode and LLVM headers somewhere? Could try removing `CPATH`.

Yup, that was it. I wonder if we're setting `CPATH` wrong. We see breakage here and there from it. Of course, we also don't see the tests that would've failed without `CPATH` but succeed because of it.

CC @Bo98 for thoughts.

**Comment by carlocab** [comment]
Yea the C++ headers have been in the SDK for a while now, at least for the CLT. Since Big Sur, I think. That's why we have

https://github.com/Homebrew/homebrew-core/blob/7fbd03f3eca777b997c24efcd71300b17ac2ec8f/Formula/llvm.rb#L243-L255

Only the `c++` path is probably needed there.

**Comment by Bo98** [comment]
> I'm guessing since the C++ headers are now in the SDK, the CPATH here is picking up some of those headers and mixing the two, probably from the #include_next <stddef.h>.

Actually thinking about it again, maybe not. We don't add the C++ directory to CPATH - only the generic SDK usr include.

It won't be a case of C++ headers being mixed, but rather the search order. The SDK usr include needs to be after LLVM's C++ directory, since it ships a `stddef.h` extension, which doesn't work the other way round.

So nothing to do with the C++ headers in the SDK at all.

**Pull request closed without merging**